### PR TITLE
[CINN] EliminateCommonGlobalVar pass, optimize performance

### PR DESCRIPTION
**Pull request by jiahy0825**
### PR types
Performance optimization

### PR changes
Others

### Description

pcard-76996

`Softmax shape=[128, 12, 128, 128]` 场景下做性能验证
phi = 248 us，CINN = 308 us，优化后 252 us，有 22.2 % 的性能提升，与 phi 基本持平

注：本 PR 的 EliminateCommonGlobalTensor pass 采取了一种较为保守的 GlobalTensor 替换策略：当且仅当同一个 GlobalTensor 在不同 ScheduleBlock 的下标索引完全相等时，才会将其替换为 LocalTensor

Perform performance validation in the `Softmax shape=[128, 12, 128, 128]` scenario
Phi=248 us, CINN=308 us, optimized to 252 us, with a 22.2% performance improvement, basically on par with phi

Note: EliminateCommonGlobalTensor pass in this PR adopts a more conservative GlobalTensor replacement strategy: it will only be replaced with LocalTensor if and only if the same GlobalTensor has exactly the same index in different ScheduleBlocks

### CODE REVIEW

File: paddle/cinn/optim/eliminate_common_global_tensor.cc [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+    if (load_buffer->memory_type == ir::MemoryType::Heap) {
+      std::vector<ir::Expr> tensor_indices;
+      for (const auto& indice : node->indices) {
+        ir::Expr new_indice = ir::ir_utils::IRCopy(indice);
+        for (const auto& [var, sb_expr] : var_to_sb_expr_) {
+          ReplaceVarWithExpr(&new_indice, var, ir::ir_utils::IRCopy(sb_expr));
+        }
+        tensor_indices.push_back(new_indice);
+      }
+      for (const auto& [var, extent] : for_var_extents_) {
```
Comment (BiynXu) [inline paddle/cinn/optim/eliminate_common_global_tensor.cc]: 这个循环体中什么也不做吗？

File: paddle/cinn/optim/eliminate_common_global_tensor.cc [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+      return true;
+    };
+
+    auto IsGlobalTensorNeedEliminate =
+        [&](const std::vector<IndicesAndExtent>& indice_and_extent) -> bool {
+      if (indice_and_extent.size() <= 1) return false;
+      return AllIndiceAndExtentEqual(indice_and_extent);
+    };
+
+    std::unordered_set<std::string> global_tensor;
```
Comment (BiynXu) [inline paddle/cinn/optim/eliminate_common_global_tensor.cc]: 最好把global_tensor都换成global_buffer_name

File: paddle/cinn/optim/eliminate_common_global_tensor.cc [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    for (const auto& [tensor_name, indice_and_extent] :
```
Comment (BiynXu) [inline paddle/cinn/optim/eliminate_common_global_tensor.cc]: tensor_name -> buffer_name 
不保证tensor_name和buffer_name全部一致，从代码上看，这个pass全部按buffer_name索引，包括对Load节点进行collect时，最好全部都用buffer_name。

File: paddle/cinn/optim/eliminate_common_global_tensor.cc [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        global_tensor_to_local_tensor_.count(buffer_name),
+        0,
+        ::common::errors::InvalidArgument(
+            "global_tensor_to_local_tensor_ should contain buffer_name %s",
+            buffer_name));
+    load_node->tensor = global_tensor_to_local_tensor_[buffer_name];
+  }
+
+  std::unordered_set<std::string> eliminate_tensor_names_;
+  std::unordered_map<std::string, ir::Expr> global_tensor_to_local_tensor_;
```
Comment (BiynXu) [inline paddle/cinn/optim/eliminate_common_global_tensor.cc]: 这个key也是buffer_name

File: paddle/cinn/optim/eliminate_common_global_tensor.cc (hunk 1 above)
Comment (jiahy0825) [reply paddle/cinn/optim/eliminate_common_global_tensor.cc]: 这个 for 循环应该删掉，thanks

File: paddle/cinn/optim/eliminate_common_global_tensor.cc (hunk 2 above)
Comment (jiahy0825) [reply paddle/cinn/optim/eliminate_common_global_tensor.cc]: Done，thanks～

File: paddle/cinn/optim/eliminate_common_global_tensor.cc (hunk 3 above)
Comment (jiahy0825) [reply paddle/cinn/optim/eliminate_common_global_tensor.cc]: Done，thanks~

File: paddle/cinn/optim/eliminate_common_global_tensor.cc (hunk 4 above)
Comment (jiahy0825) [reply paddle/cinn/optim/eliminate_common_global_tensor.cc]: Done，thanks~

**Review by tc20042008 — APPROVED** [review]

**Pull request merged**